server/src/main/java/org/elasticsearch/indices/IndicesWriteLoadStatsCollector.java

```diff
@@ -0,0 +1,242 @@
+                indexShard.shardId(),
+                shardId -> new ShardWriteLoadHistogram(
+                    parentDataStreamName,
+                    shardId,
+                    indexShard.routingEntry().primary(),
+                    relativeTimeInNanosSupplier
+                )
+            );
+            long totalIndexingTime = Math.addExact(indexShard.getTotalIndexingTimeInNanos(), indexShard.getTotalDeleteTimeInNanos());
+            long totalMergeTimeInMillis = indexShard.getTotalMergeTimeInMillis();
+            long totalRefreshTimeInNanos = indexShard.getTotalRefreshTimeInNanos();
+            shardWriteLoadHistogram.recordSample(totalIndexingTime, totalMergeTimeInMillis, totalRefreshTimeInNanos);
+        }
+    }
+
+    Iterable<IndexShard> getDataStreamsWriteIndicesShards() {
+        // Wait until the node has access to the cluster state
+        if (clusterStateReady.get() == false) {
+            return Collections.emptyList();
+        }
+
+        return () -> clusterService.state()
+            .metadata()
+            .getIndicesLookup()
+            .values()
+            .stream()
+            .filter(IndexAbstraction::isDataStreamRelated)
+            .map(dataStream -> indicesService.indexService(dataStream.getWriteIndex()))
+            .filter(Objects::nonNull)
+            .flatMap(f -> StreamSupport.stream(f.spliterator(), false))
+            .iterator();
+    }
+
+    private void cleanRolledOverIndices() {
+        final Metadata metadata = clusterService.state().metadata();
+        for (IndexAbstraction indexAbstraction : metadata.getIndicesLookup().values()) {
+            if (indexAbstraction.isDataStreamRelated() == false) {
+                continue;
+            }
+
+            Index writeIndex = indexAbstraction.getWriteIndex();
+            for (Index index : indexAbstraction.getIndices()) {
+                if (index.equals(writeIndex)) {
+                    continue;
+                }
+
+                for (IndexShard indexShard : getShardsForIndex(index)) {
+                    histograms.remove(indexShard.shardId());
+                }
+            }
+        }
+    }
+
+    protected Iterable<IndexShard> getShardsForIndex(Index index) {
+        // If the shards are allocated somewhere else, just return an empty list
+        if (indicesService.hasIndex(index) == false) {
+            return Collections.emptyList();
+        }
+
+        return indicesService.indexServiceSafe(index);
+    }
+
+    @Nullable
+    String getParentDataStreamName(String indexName) {
+        IndexAbstraction indexAbstraction = clusterService.state().getMetadata().getIndicesLookup().get(indexName);
+        IndexAbstraction.DataStream parentDataStream = indexAbstraction.getParentDataStream();
+        return parentDataStream != null ? parentDataStream.getName() : null;
+    }
+
+    public List<ShardWriteLoadDistribution> getWriteLoadDistributionAndReset() {
+        cleanRolledOverIndices();
```

Comment:
Should we prefer to collect the data from `histograms` first and then clean the rolled over indices? Just to get the last second of activity included.